docs/writing-controllers.md

```diff
@@ -0,0 +1,1447 @@
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowNotificationAction = {
+  type: `${typeof name}:showNotification`;
+  handler: () => void;
+};
+```
+
+## Use standard naming for action types
+
+Types for messenger actions should follow this naming:
+
+```
+Controller name + Action identifier without controller name prefix + "Action"
+```
+
+🚫 **Type name does not include controller name and does not end with "Action"**
+
+```typescript
+const name = 'FooController';
+
+export type ShowNotification = {
+  type: `${typeof name}:show`;
+  handler: () => void;
+};
+```
+
+🚫 **Type name includes controller name, but does not match action identifier minus prefix**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowNotificationAction = {
+  type: `${typeof name}:show`;
+  handler: () => void;
+};
+```
+
+✅ **Type name now matches controller name + action identifier + "Action"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerShowAction = {
+  type: `${typeof name}:show`;
+  handler: () => void;
+};
+```
+
+## Use standard naming for event identifiers
+
+An event is not an action, so the identifier for an event should not describe something that the controller needs to do, but rather something that has has happened to the controller.
+
+You may find it helpful to pretend that the word "on" precedes the identifier (but do not include this in the identifier itself).
+
+🚫 **`receiveMessage` sounds like a command**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerReceiveMessageEvent = {
+  type: `${typeof name}:receiveMessage`;
+  payload: () => void;
+};
+```
+
+🚫 **The event identifier includes "on"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerOnMessageReceivedEvent = {
+  type: `${typeof name}:onMessageReceived`;
+  payload: () => void;
+};
+```
+
+✅ **The event identifier sounds like a statement and does not include "on"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerMessageReceivedEvent = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+## Use standard naming for event types
+
+Types for messenger events should follow this naming:
+
+```
+Controller name + Event identifier without controller name prefix + "Event"
+```
+
+🚫 **Type name does not include controller name and does not end with "Action"**
+
+```typescript
+const name = 'FooController';
+
+export type MessageReceived = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+✅ **Type name now matches controller name + event identifier + "Event"**
+
+```typescript
+const name = 'FooController';
+
+export type FooControllerMessageReceivedEvent = {
+  type: `${typeof name}:messageReceived`;
+  payload: () => void;
+};
+```
+
+## Define and export internal action types as one
```

Comment:
The meaning of "as one" seems a little ambiguous.

```suggestion
## Define and export a type union for internal action types
```

Reply:
Makes sense.